base/parse-type.jl

```diff
@@ -0,0 +1,166 @@
+        # PERF: Drop the first element, instead of args[2:end], to avoid a new sub-vector
+        popfirst!(ast.args)
+        body = try
+            typ{ast.args...}
+        catch
+            raise && throw(ArgumentError("invalid type expression: $ast"))
+            return nothing
+        end
+        # Handle any new type vars we created
+        if !isempty(new_type_vars)
+            # Now work backwards through the new type vars and construct our wrapper UnionAlls:
+            for type_var in Iterators.reverse(new_type_vars)
+                body = UnionAll(type_var, body)
+            end
+        end
+        return body
+    elseif ast isa Expr && ast.head == :where
+        # Collect all the type vars
+        # Keep them in order, since we need to wrap the UnionAlls in reverse order.
+        new_type_vars = TypeVar[]
+        type_vars = Dict{Symbol, TypeVar}()
+        for i in 2:length(ast.args)
+            @_or_nothing type_var = _try_parse_type_var(ast.args[i], raise, type_vars)
+            type_var::TypeVar
+            type_vars[type_var.name] = type_var
+            push!(new_type_vars, type_var)
+        end
+        # Then evaluate the body in the context of those type vars
+        @_or_nothing body = _try_parse_type(ast.args[1], raise, type_vars)
+        # Now work backwards through the new type vars and construct our wrapper UnionAlls:
+        for type_var in Iterators.reverse(new_type_vars)
+            body = UnionAll(type_var, body)
+        end
+        return body
+    elseif ast isa Expr && ast.head == :call && ast.args[1] === :typeof
+        return typeof(@_or_nothing _try_parse_type(ast.args[2], raise, type_vars))
+    elseif ast isa Expr && ast.head == :call
+        return @_or_nothing _try_parse_isbits_constructor(ast, raise, type_vars)
+    else
+        return @_or_nothing _try_parse_qualified_type(ast, raise, type_vars)
+    end
+end
+_try_parse_qualified_type(val, _, _) = val
+function _try_parse_qualified_type(ast::Expr, raise::Bool, type_vars)
+    if ast.head !== :(.)
+        raise && throw(ArgumentError("Failed to parse type expression. Expected a \
+                qualified type, e.g. `Base.Dict`, got: `$ast`"))
+        return nothing
+    end
+    @_or_nothing mod = _try_parse_qualified_type(ast.args[1], raise, type_vars)
+    value = ast.args[2]
+    if value isa QuoteNode
+        value = value.value
+    end
+    return getglobal(mod, value)
+end
+function _try_parse_qualified_type(sym::Symbol, raise::Bool, type_vars)
+    # First try to look up the symbol in the type vars
+    if type_vars !== nothing
+        v_if_found = get(type_vars, sym, :not_found)
+        if v_if_found !== :not_found
+            return v_if_found
+        end
+    end
+    #@show type_vars
+    # Otherwise, look up the symbol in Main
+    if !raise  # Skip the isdefined check if !raise, since getglobal will throw anyway.
+        !isdefined(Main, sym) && return nothing
+    end
+    getglobal(Main, sym)
```

Comment:
It's not obvious to me that `Main` is the correct module here.

Wouldn't you want the ability to use the caller's module?  Maybe the module should be a parameter?

Reply:
Yeah, I wasn't sure about this either.. Thanks for reviewing this so carefully, @stevengj!

My thinking was that if the type string was printed by julia, it would be fully qualified such that any unqualified name would be parented to Main. For example:
```julia
julia> module M
       struct X1; x; end

       println(typeof(X1(2)))
       end;
Main.M.X1
```
And for types printed by julia's C static show (such as those printed via `--trace-compile`).

But it seems reasonable to support a module parameter, yeah, which would allow supporting more manually-written type strings.  Maybe I will add that as a kwarg to the parse function?

Reply:
Not everything is rooted in `Main` though.
```julia
julia> @eval Base begin
       struct X1; x; end
       println(typeof(X1(2)))
       end
Base.X1
```
and
```julia
julia> @eval Core begin
       struct X1; x; end
       println(typeof(X1(2)))
       end
Core.X1
```

Reply:
`Main.Base.X1 === Base.X1` and `Main.Core.X1 === Core.X1` though...

So I think my approach is still the best way i've seen to actually successfully find those top-level names... 🤔 do you have an alternative suggestion?

Reply:
> do you have an alternative suggestion?

You could just look for a fixedpoint of the `parentmodule` function. e.g.
```julia
julia> parentmodule(Base.Math)
Base

julia> parentmodule(ans)
Main

julia> parentmodule(ans)
Main
```
and 
```julia
julia> parentmodule(Core.Compiler)
Core

julia> parentmodule(ans)
Core
```

Reply:
ah, good idea. i think probably something there could be made to work

Reply:
Wait, no, i don't think this is right.

If we see `:(Core.Compiler)` the first time, the question is where to look up the name `:Core`. We can't look it up in `Core` because we don't have a reference to that yet! The goal is to _fetch_ that top-level name.

I guess we could look in `Base.loaded_modules()` rather than in `Main`?
Would that address your concern? That's what they do in PackageCompiler when trying to eval these kinds of strings in a module that may not have the top-level package name loaded.

Reply:
Okay, I tried that suggestion, which actually allows us to support more type strings! Here is an example test case that only works after this change:
```julia
        # Here, we know the Random module is loaded because it's an indirect dependency
        # loaded through Test. However, `Random` is not imported here, so in order for this
        # to parse, it has to look up the top-level module name in the loaded modules.
        # TODO(PR): *Should* it do this? This is maybe a bit too fancy? Would it be better
        # to just require that the module be imported to Main or the current module in order
        # to parse this?
        @assert !@isdefined(Random)
        @test parse(Type, "Random.UInt10Raw") === Test.Random.UInt10Raw
```

HOWEVER i'm not actually sure if this is a desirable feature, and I increasingly prefer @stevengj's suggestion to just provide an optional module argument that the user can specify for the context where we are parsing.

Reply:
Okay, @stevengj: I have added a parameter to pass in modules, per your suggestion, and I dropped support for looking into the loaded modules.

For now, if someone wants to parse a loaded name that isn't in-scope, they'll need to bring the module into scope (via imports or assignment).